### PR TITLE
Implement Shadow Attack

### LINKED ISSUE #295: Randomised Smoothing as Training or Certification
Investigate if Randomised Smoothing can be added as an adversarial trainer or if we should create a separate API for certifications.

### LINKED ISSUE #342: Implement Shadow Attack
Implement "Shadow Attack" as evasion attack.

Paper link: https://openreview.net/forum?id=HJxdTxHYvB

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request implements the Shadow Attack for breaking certification defences as a new evasion attack. This pull request introduces certified estimators as a new type of estimators with a first specific implementation for Randomised Smoothing for TensorFlow v2 and PyTorch providing Monte Carlo approximation of loss gradients. It removes the classifier wrapper for Randomised Smoothing.

Fixes #295 #342 

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/estimators/certification/test_randomized_smoothing.py [hunk 1]
```diff
@@ -54,24 +53,34 @@ def setUpClass(cls):
...
+        params = {"y": random_targets(y_test, ptc.nb_classes)}
         x_test_adv = fgsm.generate(x_test, **params)
 
         # Initialize RS object and attack with FGSM
-        rs = RandomizedSmoothing(classifier=krc, sample_size=100, scale=0.01, alpha=0.001)
+        rs = PyTorchRandomizedSmoothing(model=ptc.model,
+                                        loss=ptc._loss,
+                                        input_shape=ptc.input_shape,
+                                        nb_classes=ptc.nb_classes,
+                                        channel_index=ptc.channel_index,
```
Comment (hesseltuinhof) [inline tests/estimators/certification/test_randomized_smoothing.py:79]: See previous comments.

File: tests/estimators/certification/test_randomized_smoothing.py [hunk 2]
```diff
@@ -117,11 +118,19 @@ def setUp(self):
...
         (_, _), (x_test, y_test) = self.iris
 
-        krc = get_tabular_classifier_kr()
-        rs = RandomizedSmoothing(classifier=krc, sample_size=100, scale=0.01, alpha=0.001)
+        ptc = get_tabular_classifier_pt()
+        rs = PyTorchRandomizedSmoothing(model=ptc.model,
+                                        loss=ptc._loss,
+                                        input_shape=ptc.input_shape,
+                                        nb_classes=ptc.nb_classes,
+                                        channel_index=ptc.channel_index,
```
Comment (hesseltuinhof) [inline tests/estimators/certification/test_randomized_smoothing.py:126]: See previous comments.

File: notebooks/output_randomized_smoothing_mnist.ipynb [hunk 3]
```diff
@@ -109,128 +112,115 @@
...
     "\n",
-    "# sigma = 0.25\n",
-    "ga = GaussianAugmentation(sigma=sigma1, augmentation=False)\n",
-    "x_new1, _ = ga(x_train)\n",
+    "classifier = TensorFlowV2Classifier(model=get_model(input_shape, min_, max_),\n",
+    "                                    nb_classes=nb_classes,\n",
+    "                                    input_shape=input_shape,\n",
+    "                                    loss_object=loss_object,\n",
+    "                                    train_step=train_step,\n",
+    "                                    channel_index=3,\n",
```
Comment (hesseltuinhof) [inline notebooks/output_randomized_smoothing_mnist.ipynb:159]: Replace `channel_index` in favor of `channels_first`.

File: notebooks/output_randomized_smoothing_mnist.ipynb [hunk 4]
```diff
@@ -109,128 +112,115 @@
...
+    "sigma_0 = 0.5\n",
     "\n",
-    "classifier_ga2 = cnn_mnist(x_train.shape[1:], min_, max_)\n",
-    "classifier_ga2.fit(x_new2, y_train, nb_epochs=num_epochs, batch_size=128)"
+    "classifier_rs_0 = TensorFlowV2RandomizedSmoothing(model=classifier.model,\n",
+    "                                                  nb_classes=nb_classes,\n",
+    "                                                  input_shape=input_shape,\n",
+    "                                                  loss_object=loss_object,\n",
+    "                                                  train_step=train_step,\n",
+    "                                                  channel_index=3,\n",
```
Comment (hesseltuinhof) [inline notebooks/output_randomized_smoothing_mnist.ipynb:171]: See the previous comment.

File: notebooks/output_randomized_smoothing_mnist.ipynb [hunk 5]
```diff
@@ -109,128 +112,115 @@
...
+    "# Create and train smoothed classifier, sigma = 0.25\n",
+    "\n",
+    "sigma_1 = 0.25\n",
+    "\n",
+    "classifier_rs_1 = TensorFlowV2RandomizedSmoothing(model=get_model(input_shape, min_, max_),\n",
+    "                                                 nb_classes=nb_classes,\n",
+    "                                                 input_shape=input_shape,\n",
+    "                                                 loss_object=loss_object,\n",
+    "                                                 train_step=train_step,\n",
+    "                                                 channel_index=3,\n",
```
Comment (hesseltuinhof) [inline notebooks/output_randomized_smoothing_mnist.ipynb:193]: See the previous comment.

File: notebooks/output_randomized_smoothing_mnist.ipynb [hunk 6]
```diff
@@ -109,128 +112,115 @@
...
+    "# Create and train smoothed classifier, sigma = 0.5\n",
+    "\n",
+    "sigma_2 = 0.5\n",
+    "\n",
+    "classifier_rs_2 = TensorFlowV2RandomizedSmoothing(model=get_model(input_shape, min_, max_),\n",
+    "                                                  nb_classes=nb_classes,\n",
+    "                                                  input_shape=input_shape,\n",
+    "                                                  loss_object=loss_object,\n",
+    "                                                  train_step=train_step,\n",
+    "                                                  channel_index=3,\n",
```
Comment (hesseltuinhof) [inline notebooks/output_randomized_smoothing_mnist.ipynb:217]: See the previous comment.

File: art/attacks/evasion/shadow_attack.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+
+    def __init__(
+        self,
+        estimator,
+        sigma=0.5,
+        nb_steps=300,
+        learning_rate=0.1,
+        lambda_tv=0.1,
+        lambda_c=20.0,
+        lambda_s=10.0,
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/shadow_attack.py:69]: I think these were the values for one of the ImageNet experiments. I have now set these default values according to the CIFAR-10 experiments in Table 1.

File: art/attacks/evasion/shadow_attack.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+                raise ValueError("Target labels `y` need to be provided for a targeted attack.")
+
+            logger.info("Using model predictions as correct labels for FGM.")
+            y = get_labels_np_array(self.estimator.predict(x, batch_size=self.batch_size))
+        else:
+            self.targeted = True
+
+        if x.ndim != 4:
+            raise ValueError(
+                "Feature vectors detected. The adversarial patch can only be applied to image data dimensions."
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/shadow_attack.py:144]: I like it, we should use it where possible.

File: docs/index.rst [hunk 9]
```diff
@@ -145,6 +145,8 @@ Implemented Attacks, Defences, Detections, Metrics, Certifications and Verificat
    modules/defences/transformer
    modules/detection
    modules/estimators
+   modules/estimators/certification
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply docs/index.rst:148]: Yes, that's ok. A PR adding a new module can also include a related update of the docs.

File: art/attacks/evasion/shadow_attack.py [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+        :param perturbation: The perturbation to be regularised.
+        :type perturbation: `np.ndarray`
+        :return: The loss gradients of the perturbation.
+        """
+
+        if self.frame_work == "tensorflow":
+
+            import tensorflow as tf
+
+            if self.estimator.channel_index == 3:
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/shadow_attack.py:196]: Ok.

File: art/attacks/evasion/shadow_attack.py [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+                            axis=(1, 2),
+                        )
+
+                    loss_c = tf.norm(tf.reduce_mean(tf.abs(perturbation_t), axis=[2, 3]), ord=2, axis=1) ** 2
+
+                    loss = self.lambda_tv * loss_tv + self.lambda_s * loss_s + self.lambda_c * loss_c
+
+                    gradients = tape.gradient(loss, perturbation_t).numpy()
+
+                    if self.estimator.channel_index == 3:
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/shadow_attack.py:227]: Ok.

File: art/attacks/evasion/shadow_attack.py [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+                        gradients = gradients.transpose(0, 2, 3, 1)
+
+            else:
+                raise ValueError("Expecting eager execution.")
+
+        elif self.frame_work == "pytorch":
+
+            import torch
+
+            perturbation_t = torch.from_numpy(perturbation).to("cpu")
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/shadow_attack.py:237]: My thought was that the calculation of the regularisation loss will always be a small computation and it would never get any advantage from running on a GPU.

File: art/attacks/evasion/shadow_attack.py [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+        if not isinstance(self.learning_rate, float):
+            raise ValueError("The learning rate must be of type float.")
+
+        if self.learning_rate <= 0:
+            raise ValueError("The learning rate must larger than zero.")
+
+        if not isinstance(self.lambda_tv, float):
+            raise ValueError("The lambda_tv must be of type float.")
+
+        if self.lambda_tv <= 0:
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/shadow_attack.py:294]: Good idea.

File: art/attacks/evasion/shadow_attack.py [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+            raise ValueError("The lambda_tv must larger than zero.")
+
+        if not isinstance(self.lambda_c, float):
+            raise ValueError("The lambda_c must be of type float.")
+
+        if self.lambda_c <= 0:
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/shadow_attack.py:300]: Good idea.

File: art/attacks/evasion/shadow_attack.py [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+            raise ValueError("The lambda_c must larger than zero.")
+
+        if not isinstance(self.lambda_s, float):
+            raise ValueError("The lambda_s must be of type float.")
+
+        if self.lambda_s <= 0:
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/shadow_attack.py:306]: Good idea.

File: art/estimators/certification/randomized_smoothing/pytorch.py [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+    """
+
+    def __init__(
+        self,
+        model,
+        loss,
+        input_shape,
+        nb_classes,
+        optimizer=None,
+        channel_index=1,
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/estimators/certification/randomized_smoothing/pytorch.py:50]: Yes.

File: art/estimators/certification/randomized_smoothing/tensorflow.py [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+    """
+
+    def __init__(
+        self,
+        model,
+        nb_classes,
+        input_shape,
+        loss_object=None,
+        train_step=None,
+        channel_index=3,
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/estimators/certification/randomized_smoothing/tensorflow.py:48]: Ok.

File: art/estimators/certification/randomized_smoothing/tensorflow.py [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+        :param alpha: The failure probability of smoothing
+        :type alpha: `float`
+        """
+        super().__init__(
+            model=model,
+            nb_classes=nb_classes,
+            input_shape=input_shape,
+            loss_object=loss_object,
+            train_step=train_step,
+            channel_index=channel_index,
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/estimators/certification/randomized_smoothing/tensorflow.py:99]: Ok.

File: tests/estimators/certification/test_randomized_smoothing.py (hunk 1 above)
Comment (beat-buesser) [reply tests/estimators/certification/test_randomized_smoothing.py:79]: Ok.

File: tests/estimators/certification/test_randomized_smoothing.py (hunk 2 above)
Comment (beat-buesser) [reply tests/estimators/certification/test_randomized_smoothing.py:126]: Ok.

File: art/estimators/certification/randomized_smoothing/pytorch.py [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+            postprocessing_defences=postprocessing_defences,
+            preprocessing=preprocessing,
+            device_type=device_type,
+            sample_size=sample_size,
+            scale=scale,
+            alpha=alpha,
+        )
+
+    def _predict_classifier(self, x, batch_size):
+        x = x.astype(np.float32)
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/estimators/certification/randomized_smoothing/pytorch.py:117]: Ok.

File: art/estimators/certification/randomized_smoothing/pytorch.py [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+        return PyTorchClassifier.predict(self, x=x, batch_size=batch_size)
+
+    def _fit_classifier(self, x, y, batch_size, nb_epochs, **kwargs):
+        x = x.astype(np.float32)
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/estimators/certification/randomized_smoothing/pytorch.py:121]: Ok.

File: art/estimators/certification/randomized_smoothing/randomized_smoothing.py [hunk 21]
```diff
@@ -101,62 +117,74 @@ def predict(self, x, batch_size=128, is_abstain=True, **kwargs):
             print("%s prediction(s) abstained." % n_abstained)
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/estimators/certification/randomized_smoothing/randomized_smoothing.py:117]: Yes.

File: notebooks/output_randomized_smoothing_mnist.ipynb (hunk 3 above)
Comment (beat-buesser) [reply notebooks/output_randomized_smoothing_mnist.ipynb:159]: Ok.

File: notebooks/output_randomized_smoothing_mnist.ipynb (hunk 4 above)
Comment (beat-buesser) [reply notebooks/output_randomized_smoothing_mnist.ipynb:171]: Ok.

File: notebooks/output_randomized_smoothing_mnist.ipynb (hunk 5 above)
Comment (beat-buesser) [reply notebooks/output_randomized_smoothing_mnist.ipynb:193]: Ok.

File: notebooks/output_randomized_smoothing_mnist.ipynb (hunk 6 above)
Comment (beat-buesser) [reply notebooks/output_randomized_smoothing_mnist.ipynb:217]: Ok.
